crates/red_knot_python_semantic/src/semantic_index/narrowing_constraints.rs

```diff
@@ -0,0 +1,162 @@
+//! # Narrowing constraints
+//!
+//! When building a semantic index for a file, we associate each binding with _narrowing
+//! constraints_. The narrowing constraint is used to constrain the type of the binding's symbol.
+//! Note that a binding can be associated with a different narrowing constraint at different points
+//! in a file. See the [`use_def`][crate::semantic_index::use_def] module for more details.
+//!
+//! This module defines how narrowing constraints are stored internally.
+//!
+//! A _narrowing constraint_ consists of a list of _clauses_, each of which corresponds with an
+//! expression in the source file (represented by a [`Constraint`]). We need to support the
+//! following operations on narrowing constraints:
+//!
+//! - Adding a new clause to an existing constraint
+//! - Merging two constraints together, which produces the _intersection_ of their clauses
+//! - Iterating through the clauses in a constraint
+//!
+//! In particular, note that we do not need random access to the clauses in a constraint. That
+//! means that we can use a simple [_sorted association list_][ruff_index::list] as our data
+//! structure.
+//!
+//! [`Constraint`]: crate::semantic_index::constraint::Constraint
+
+use ruff_index::list::{ListBuilder, ListIterator, ListStorage};
+use ruff_index::newtype_index;
+
+use crate::semantic_index::constraint::ScopedConstraintId;
+
+/// A narrowing constraint associated with a live binding.
+///
+/// A constraint is a list of clauses, each of which is a [`Constraint`] that constrains the type
+/// of the binding's symbol.
+///
+/// An instance of this type represents a _non-empty_ narrowing constraint. You will often wrap
+/// this in `Option` and use `None` to represent an empty narrowing constraint.
+///
+/// [`Constraint`]: crate::semantic_index::constraint::Constraint
+#[newtype_index]
+pub(crate) struct ScopedNarrowingConstraintId;
+
+/// One of the clauses in a narrowing constraint, which is a [`Constraint`] that constrains the
+/// type of the binding's symbol.
+///
+/// Note that those [`Constraint`]s are stored in [their own
+/// arena][crate::semantic_index::constraint::Constraints], so internally we use a
+/// [`ScopedConstraintId`] to refer to the underlying constraint.
+///
+/// [`Constraint`]: crate::semantic_index::constraint::Constraint
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub(crate) struct ScopedNarrowingConstraintClause(ScopedConstraintId);
```

Comment:
This might seem like an unnecessary extra wrapper, but I have hopes to move `is_positive` from `Constraint` to here, so I wanted to go ahead and have this type available as the place to move it to.  Also I think it adds clarity to the distinction between _constraints_ (the expressions in the source code), and _narrowing/visibility constraints_ (the two ways in which we're using those).